### ISSUE TITLE
docker treats example.com:443 and example.com as different registries

### ISSUE THREAD

**Issue by vikstrous** [docker/docker#22644]
Docker has always treated `example.com/user/repo` and `example.com:443/user/repo` as separate repos. This has been a source of errors and confusion with DTR users. This is a problem with Notary because Notary signs the repo name including the domain. There is no way to transform signed Notary data from one format to the other. This means that if you sign an image with the wrong one, you have to forever make sure no one switches their client to use the other format.

I propose that we standardize on which one of these is the correct name. I don't care which one it is, it just needs to be consistent.

One way to do this is to internally treat `example.com/user/repo` as `example.com:443/user/repo`. Let me know if you guys have other ideas for how to resolve this.

**Comment by thaJeztah** [comment]
My guess is that the thought behind this is that multiple registries can run on the same domain, but using different ports (example.com:5000, example.com:5001).

**Comment by vikstrous** [comment]
Yeah, and that's okay. The issue is that port 443 is special because you can have two different tags in docker that are actually the same registry, namespace, repo, and tag, just with or without port 443 being explicit.

**Comment by thaJeztah** [comment]
But without a protocol, would it be `example.com:80` or `example.com:443`?

**Comment by vikstrous** [comment]
Hmm... That's a good question. I guess it's both right now. That's very confusing. In DTR's case we never allow non-HTTPS connections so it's always 443, but in general --insecure-registry falls back to both. This is just really deeply baked into the design of docker image names and it'll continue to cause problems in unexpected places :( Ideally, I don't think that domain name should be part of the image name at all, but that's a different discussion. I don't know what we can do here.

**Comment by thaJeztah** [comment]
ping @stevvooe @dmcgowan any suggestions?

**Comment by dmcgowan** [comment]
We are working on a proposal for always sending up the hostname component of the repository name to the registry. It's design is primarily for proxies but also discussing other uses for it on the backend. From a naming perspective they are different images and will remain different images, and I agree it would be ideal for the registry to recognize them as such. For login, I don't think it would be unreasonable to treat the two names the same, this would only effect how to the client determines the auth config.

When I have a proposal created I will link it here.

**Comment by stevvooe** [comment]
The simple fact here is that `example.com` and `example.com:443` are different hosts. The former will switch ports based on protocol (`http` or `https`, or other future additions) while the latter will always use port 443, even when `http` or other protocol is specified.

This can be confirmed by running a plaintext server on port 443. The following works, because we are port bound:

```console
$ curl -v http://localhost:443
```

Even though we are running on 443, because the host is port bound, this works because the scheme `http` will be used. With the same plaintext server, the following doesn't work:

```console
$ curl -v https://localhost/
*   Trying ::1...
* Connected to localhost (::1) port 443 (#0)
* Unknown SSL protocol error in connection to localhost:-9847
* Closing connection 0
curl: (35) Unknown SSL protocol error in connection to localhost:-9847
```

Even though we are running on port 443, since the server is plaintext and the scheme is `https`, TLS will be used. The plaintext server obviously fails here. Another way of putting this is that the scheme is what selects TLS, not the port.

Any attempt to canonicalize these names would end up binding the naming too tightly to http/https. I also suspect that specifying the port as 443 may open a client up to potential security problems, since the assumption is that this port is secure, when it is clear that if the host is port bound, a plaintext connection can be made over this port. Departing from the conventions that these are different hosts in http and tls may open us up to a number of other issues.

The correct solution here is to remove the impetus to include the port in image names by decoupling the registry location from the name of the image. The best practice is to use `example.com` over `example.com:443`.

**Comment by vikstrous** [comment]
I agree completely. The port should never have been a part of the image name. I think the sanest way to manage the locations of registries is to follow the same model as git remotes. This would be fundamental change to how custom registries are handled, but I think it's necessary to reduce confusion and all the obscure bugs that stem from the image naming convention. Is there any existing work in this direction?

**Comment by Vish0007** [comment]
I have one registry but it exposes different ports, like this registry:8000, registry:8001, registry:8002 and so on. I do this to keep multiple credentials of the registry on  the docker client auth file (otherwise it keeps only on user credentials). my --insecure-registry flag is "registry/0"
I have this working for long time. I donot see any issues with that.

**Comment by stevvooe** [comment]
@Vish0007 Why are you running the registry on several different ports?

In general, it's best to access a registry through a single host name.

**Comment by Vish0007** [comment]
i want to store different more than one user credential on docker client side, so i work around it with several ports for same registry.
Btw, why its best to access registry with only one hostname (:port)? Are there any synchronization issues when we use different ports to access the registry? for example, executing docker push in parallel from two different docker clients: docker push registry:8000 and docker push 8001.
